Speed up the Gradle build

snazy opened the pull request:
There are a few tweaks in this change:
* The most "prominent" one is that the Gradle build's now called from "its" parent pom, only the "Maven" parts (install to local repo, etc) are "left" in the submodules. Gradle tests are also run from that parent pom. This equates to how a "native" Gradle build would work.
* The Gradle daemon is no longer terminated synchronously, but use a daemon idle timeout of 10s, so that the now separate `./gradlew clean` followed by the `./gradlew build/assemble` share the running daemon.

Following (non scientific) measurements from my local machine using `mvnd clean install -Dquickly`.

Before this change:
```
[INFO] Quarkus - Gradle Plugin - Parent ................... SUCCESS [  0.021 s]
[INFO] Quarkus - Gradle Model ............................. SUCCESS [ 10.175 s]
[INFO] Quarkus - Gradle Plugin ............................ SUCCESS [ 12.984 s]
[INFO] Quarkus - Extension Gradle Plugin .................. SUCCESS [ 15.500 s]
```

After this change:
```
[INFO] Quarkus - Gradle Plugin - Parent ................... SUCCESS [  6.108 s]
[INFO] Quarkus - Gradle Model ............................. SUCCESS [  0.216 s]
[INFO] Quarkus - Gradle Plugin ............................ SUCCESS [  0.052 s]
[INFO] Quarkus - Extension Gradle Plugin .................. SUCCESS [  0.025 s]
```

snazy committed:
Speed up the Gradle build

There are a few tweaks in this change:
* The most "prominent" one is that the Gradle build's now called from
  "its" parent pom, only the "Maven" parts (install to local repo, etc)
  are "left" in the submodules. Gradle tests are also run from that
  parent pom. This equates to how a "native" Gradle build would work.
* The Gradle daemon is no longer terminated synchronously, but use a
  daemon idle timeout of 10s, so that the now separate `./gradlew clean`
  followed by the `./gradlew build/assemble` share the running daemon.

Following (non scientific) measurements from my local machine using
`mvnd clean install -Dquickly`.

Before this change:
```
[INFO] Quarkus - Gradle Plugin - Parent ................... SUCCESS [  0.021 s]
[INFO] Quarkus - Gradle Model ............................. SUCCESS [ 10.175 s]
[INFO] Quarkus - Gradle Plugin ............................ SUCCESS [ 12.984 s]
[INFO] Quarkus - Extension Gradle Plugin .................. SUCCESS [ 15.500 s]
```

After this change:
```
[INFO] Quarkus - Gradle Plugin - Parent ................... SUCCESS [  2.377 s]
[INFO] Quarkus - Gradle Model ............................. SUCCESS [  0.216 s]
[INFO] Quarkus - Gradle Plugin ............................ SUCCESS [  0.052 s]
[INFO] Quarkus - Extension Gradle Plugin .................. SUCCESS [  0.025 s]
```

snazy commented:
(Updated the description to reflect the time to start the Gradle daemon)

famod commented on devtools/gradle/pom.xml:
```diff
@@ -46,14 +46,42 @@
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-clean-plugin</artifactId>
```
Shouldn't this be 
```xml
<groupId>org.codehaus.mojo</groupId>
<artifactId>exec-maven-plugin</artifactId>
```
instead?
clean plugin doesn't support an `executable`.

snazy replied:
oops - left-over. removed it.

famod commented:
Confirmed a massive speedup locally! 👍 

But neither of the following three executions run a gradle build anymore:
- `cd devtools/gradle/gradle-model && mvn -Dquickly`
- `mvn -Dquickly -pl devtools/gradle/gradle-model`
- `mvn -Dquickly -f devtools/gradle/gradle-model`

The other sibling modules are surely affected as well.

famod commented on devtools/gradle/pom.xml:
```diff
@@ -62,19 +90,29 @@
                                         <argument>-Dmaven.repo.local=${settings.localRepository}</argument>
-                                        <argument>-S</argument>
                                         <argument>--stacktrace</argument>
-                                        <argument>--no-daemon</argument>
```
This was added way back via #11694. I'm not saying it's still 100% relevant, but IMO it would be better to not have the daemon around all the time, especially in CI?

/cc @jaikiran

snazy replied:
Updated the PR a little.
Local builds --> with daemon
CI builds --> no daemon

famod replied:
Those 10s should be alright for CI. Better have as little differences as possible between local and CI, no?

Do you have an opinion on this @gsmet?

snazy committed:
remove left-over/unnecessary profile

snazy commented:
> But neither of the following three executions run a gradle build anymore:

Yes, that's the downside of this approach. Not sure whether it's relevant though, when you work on the Gradle code.

snazy committed:
No Gradle daemon in CI

snazy committed:
Remove 10s Gradle daemon timeout (CI uses no daemon)

glefloch reviewed (approved):
LGTM thanks !

famod commented:
> > But neither of the following three executions run a gradle build anymore:
> 
> Yes, that's the downside of this approach. Not sure whether it's relevant though, when you work on the Gradle code.



> > But neither of the following three executions run a gradle build anymore:
> 
> Yes, that's the downside of this approach. Not sure whether it's relevant though, when you work on the Gradle code.

There is a risk of people/conteibutors not knowing this "speciality" but if nobody else complains now we could try.

*But I do see a potential issue with incremental CI.*
I might have some time on Sunday to check that.

famod commented:
> **But I do see a potential issue with incremental CI.** I might have some time on Sunday to check that.

I'm afraid this approach won't work in incremental CI (PR builds):
If e.g. `devtools/gradle/gradle-application-plugin` contains changes (or is affected by upstream changes) then only this very module is built, not its parent. So in such a case and with this PR, we would loose test executions in the "JVM Tests" jobs.
Same applies to the other modules under `devtools/gradle`.

So we really need to retain the gradle build when using `mvn -pl ...`.

Another solution could be a special setup only for `quickly` (which isn't used in CI, at least last time I checked).
Such bigger differences betwenn local builds and CI builds can easily cause confusions, though.

snazy commented:
Guess we could keep a single maven profile and have the „right“ activation conditions. With a small Gradle daemon timeout all builds should benefit.

gsmet commented:
What's the status of this? Should we merge?

famod reviewed (changes requested):
I hate to say it but I think we should _not_ merge it as is because ([self quote](https://github.com/quarkusio/quarkus/pull/33144#issuecomment-1537533418)):
> I'm afraid this approach won't work in incremental CI (PR builds)

snazy committed:
Enable incremental CI

If the environment variable `CI_INCREMENTAL` (added to the CI job) is set, build Gradle projects incrementally. Otherwise use the "combined build" from the parent pom.

The Gradle daemon is enabled by default, with an idle timeout of 10s in all CI builds.

snazy commented:
Added another commit that treats builds differently:
* Incremental CI builds: builds per-project, not from the parent pom
* "Full" CI and other builds: builds from the parent pom

gsmet commented:
I will close this PR for lack of activity.
